core/server/src/main/java/alluxio/worker/block/evictor/LIRSEvictor.java

```diff
@@ -0,0 +1,573 @@
+        BlockStoreLocation location =
+            new BlockStoreLocation(tierView.getTierViewAlias(), dirView.getDirViewIndex());
+        Pair<BlockStoreLocation, Long> blockEntry =
+            new Pair<BlockStoreLocation, Long>(location, blockId);
+        mHIRCache.remove(blockEntry);
+        mLIRCache.remove(blockEntry);
+        mTmpCache.remove(blockEntry);
+      }
+    }
+  }
+
+  private void updateOnAccess(long blockId) {
+    for (StorageTierView tier : mManagerView.getTierViews()) {
+      for (StorageDirView dir : tier.getDirViews()) {
+        BlockStoreLocation location =
+            new BlockStoreLocation(tier.getTierViewAlias(), dir.getDirViewIndex());
+        Pair<BlockStoreLocation, Long> key = new Pair<BlockStoreLocation, Long>(location, blockId);
+        SpaceContainer spaceContainer = mSpaceManager.get(location);
+        long blockSize = mBlockIdToSize.get(blockId);
+        // 1: Block hits in tmp cache
+        if (mTmpCache.containsKey(key)) {
+          mTmpCache.remove(key);
+          // Move the block to LIR cache if a ghost block exists in LIR cache
+          if (mLIRCache.containsKey(key)) {
+            BlockLIRSInfo blockinfo = mLIRCache.get(key);
+            spaceContainer.incrementLIR(blockSize);
+            blockinfo.setIsLIR(true);
+            blockinfo.setResident(true);
+            mLIRCache.remove(key);
+            mLIRCache.put(key, blockinfo);
+          } else if (spaceContainer.getLIRBytes() + blockSize > spaceContainer.getCapacity()
+              * mLIRPercent) {
+            // Move the block to HIR cache if LIR cache is full
+            spaceContainer.incrementHIR(blockSize);
+            BlockLIRSInfo blockinfo = new BlockLIRSInfo(false, true);
+            mHIRCache.put(key, blockinfo);
+            mLIRCache.put(key, blockinfo);
+          } else {
+            // Move the block to LIR cache if LIR cache is not full
+            spaceContainer.incrementLIR(blockSize);
+            BlockLIRSInfo blockinfo = new BlockLIRSInfo(true, true);
+            mLIRCache.put(key, blockinfo);
+          }
+        } else if (mHIRCache.containsKey(key)) {
+          // 2. Block hits in HIR cache
+          BlockLIRSInfo blockinfo = mHIRCache.get(key);
+          if (mLIRCache.containsKey(key)) {
+            // Move the block to LIR cache if a ghost block exists in LIR cache
+            blockinfo.setIsLIR(true);
+            blockinfo.setResident(true);
+            mHIRCache.remove(key);
+            mLIRCache.remove(key);
+            mLIRCache.put(key, blockinfo);
+            spaceContainer.moveBlockFromHIRToLIR(blockSize);
+          } else {
+            // Move the block to HIR cache if no ghost block exists
+            mHIRCache.remove(key);
+            mHIRCache.put(key, blockinfo);
+            mLIRCache.put(key, blockinfo);
+          }
+        } else if (mLIRCache.containsKey(key)) {
+          // 3. Block hits in LIR cache, just move the block to top of LIR cache

```

Comment:
LIR cache and HIR cache are implemented with 2 LRU stacks: LIR stack and HIR stack. HIR stack only stores resident HIR blocks, while LIR stack holds all HIR blocks, LIR blocks and ghost blocks whose recency is smaller than the bottom block in LIR stack. The bottom block in LIR stack is guaranteed to be a LIR block with maximum IRR in LIR cache. Therefore, when a HIR block or ghost block is reference and the block is in LIR stack, the new IRR of the block will be lower than  the maximum IRR in LIR cache. So the block will replace the block with maximum IRR in LIR cache to be a new LIR block. For more detail info, reference to [LIRS paper](http://www.openu.ac.il/home/wiseman/2os/lru/lirs.pdf)